src/contexts/EditorDOMEventRegistryContext.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  DependencyList,
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+} from "react";
+
+import { useForceUpdate } from "../hooks/useForceUpdate";
+import {
+  DOMEventMap,
+  EventHandler,
+  createReactEventPlugin,
+} from "../plugins/createReactEventPlugin";
+
+/**
+ * Produces a plugin that can be used with ProseMirror to handle DOM
+ * events at the EditorView.dom element.
+ *
+ * - `reactEventsPlugin` is a ProseMirror plugin for handling DOM events
+ * at the EditorView.dom element. It should be passed to `useEditorView`,
+ * along with any other plugins.
+ *
+ * - 'registry' should be passed as the value to
+ * `EditorDOMEventRegistryContext`.
+ */
+export function useReactEventPlugin() {
+  const registry = useRef(
+    new Map<keyof DOMEventMap, Set<EventHandler>>()
+  ).current;
+
+  const forceUpdate = useForceUpdate();
+
+  const register = useCallback(
+    (eventType: keyof DOMEventMap, handler: EventHandler) => {
+      const handlers = registry.get(eventType) ?? new Set<EventHandler>();
+      handlers.add(handler);
+      if (!registry.has(eventType)) {
+        // We only need to force an update when the set of event types
+        // that we care about changes
+        registry.set(eventType, handlers);
+        forceUpdate();
+      }
+    },
+    [registry, forceUpdate]
+  );
+
+  const unregister = useCallback(
+    (eventType: keyof DOMEventMap, handler: EventHandler) => {
+      const handlers = registry.get(eventType);
+      if (!handlers) return;
+      handlers.delete(handler);
+    },
+    [registry]
+  );
+
+  const reactEventPlugin = useMemo(
+    () => createReactEventPlugin(registry),
+    // We force an update whenever the set of event types that
+    // we care about changes. This plugin only needs to be
+    // recreated when that set (the set of registry keys)
+    // changes.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [registry.keys()]
+  );
+
+  const editorDOMEventRegistry = useMemo(
+    () => ({ register, unregister }),
+    [register, unregister]
+  );
+
+  return { editorDOMEventRegistry, reactEventPlugin };
+}
+
+type Register<EventType extends keyof DOMEventMap = any> = (
```

Comment:
Do you need a default for the template parameter?

Reply:
I notice you don't have it on `useEditorDOMEvent`.

Reply:
Basically, yes, I _think_. Or at least, I do at _some_ point, because at some point I have to pass this type to `createContext`, and I don't think I can have the result of `createContext` be a parameterized type?

I don't need it on `useEditorDOMEvent` because it can be inferred from usage (that's the whole goal, really, is to support type inference of the `event` parameter based on the `eventType`)

Reply:
Is there any value in using keyof DOMEventMap here, or should we just use `string`?

Reply:
Yes; the value is that we can correlate the types, so that the event parameter has the correct type! For example, in the demo in this branch, because the eventtype is "keydown", the `event` is typed as `KeyboardEvent`. This is the same type correlation that ProseMirror does for `handleDOMEvents`

Reply:
Ah, okay. Can we use `= keyof DOMEventMap` rather than `= any` then?

Reply:
No :P there's some covariance/contravariance thing that I'm having trouble wrapping my head around, but if I default to keyof DOMEventMap, then I get an error at the actual callsite in the `useEditorDOMEvent` hook that EventType could be instantiated with a different subtype

Reply:
I think maybe what's happening here is that you have a generic type whose instantiations are non-generic functions rather than generic function types. I don't know if that actually makes sense, but am I wrong or does this seem to work?

```diff
diff --git a/src/contexts/EditorDOMEventRegistryContext.tsx b/src/contexts/EditorDOMEventRegistryContext.tsx
index f69d975296c4..5ace8df2e0f8 100644
--- a/src/contexts/EditorDOMEventRegistryContext.tsx
+++ b/src/contexts/EditorDOMEventRegistryContext.tsx
@@ -74,18 +74,16 @@ export function useReactEventPlugin() {
   return { editorDOMEventRegistry, reactEventPlugin };
 }
 
-type Register<EventType extends keyof DOMEventMap = any> = (
-  eventType: EventType,
-  handler: EventHandler<EventType>
-) => void;
-
-type Unregister<EventType extends keyof DOMEventMap = any> =
-  Register<EventType>;
-
-type EditorDOMEventRegistryContextValue = {
-  register: Register;
-  unregister: Unregister;
-};
+interface EditorDOMEventRegistryContextValue {
+  register<EventType extends keyof DOMEventMap>(
+    eventType: EventType,
+    handler: EventHandler<EventType>
+  ): void;
+  unregister<EventType extends keyof DOMEventMap>(
+    eventType: EventType,
+    handler: EventHandler<EventType>
+  ): void;
+}
 
 const EditorDOMEventRegistryContext = createContext(
   null as unknown as EditorDOMEventRegistryContextValue
```

Reply:
Ohhhhhhhh... yes I bet this will work. Gonna try it. Ugh I should have known I literally just had to do this in a different project this morning!!!